Bug/mid flow onboarding dropouts

ThanhDatVu111 opened the pull request:
## 🚀 What does this PR do?

> Separates user signup and onboarding logic in both the frontend and backend. 
> Ensures users who abandon the onboarding process are redirected to complete it upon returning, while first-time users can proceed directly without unnecessary API calls.

---

## ✅ Checklist

- [x] Code compiles & runs
- [x] Lint checks pass
- [ ] Updated `.env.example` if needed
- [x] PR linked to related issue/ticket (if applicable)

---

## 🔍 How to test this

### Frontend
1. **Signup Flow:**
   - Create a new user and verify that their information is saved immediately after signup.
   - Ensure the user is redirected to the onboarding flow.

2. **Onboarding Abandonment:**
   - Partially complete the onboarding process, then refresh or close the app.
   - Verify that the user is redirected back to the onboarding flow upon returning.

3. **Returning Users:**
   - Complete the onboarding process, then log out and log back in.
   - Verify that the user is redirected to the home page.

### Backend
1. **API Testing:**
   - Test `/signup`, `/onboard`, and `/fetch` endpoints using curl.
   - Verify that the `is_onboarding_complete` flag is updated correctly in the database.

---

## 🧠 Extra Notes

### Frontend
- Refactored `AuthRoutesLayout` to handle user redirection based on onboarding status.
- Added logic to detect incomplete onboarding and redirect users accordingly.
- Separated API calls for user signup (`createUser`) and onboarding (`onboardUser`).
- Updated `fetchUser` API to fetch user data independently of the onboarding flow.

### Backend
- Added `/signup`, `/onboard`, and `/fetch` endpoints in the backend.
- Updated `userModel` and `userService` to handle user creation, updates, and fetching.
- Added `is_onboarding_complete` logic for tracking onboarding status.
- Updated database schema for `users` table to include the `is_onboarding_complete` flag.

> - For the layout file, I implemented logic to handle cases where users abandon the onboarding process. When they return, the app fetches their status and redirects them to the correct page.
> - But For first-time users, the app still unnecessarily fetches the user data when they just signup - I will try to figure this out.

ThanhDatVu111 committed:
feat(tabs): add bottom tab navigator with Home, Library, Calendar, Chat, and Profile

ThanhDatVu111 committed:
fixed  styling for tabs and added some minor changes

ThanhDatVu111 committed:
Add tab-screen Tailwind plugin, bump Expo versions & refactor screens

- Introduce a custom  component via TailwindCSS plugin
- Upgrade Expo SDK dependencies (expo, expo-router, expo-splash-screen) to expected patch versions
- Refactor Chat, Calendar, Message & Home screens to use the new shared CSS utility for cleaner markup

ThanhDatVu111 committed:
change package json lock in backend

ThanhDatVu111 committed:
feat(frontend): decouple signup & onboarding flows

- Split createUser & onboardUser APIs
- Update fetchUser outside of onboarding
- Redirect users to pending onboarding on return
- Persist signup data immediately

Ensure incomplete onboarding resumes correctly.

ThanhDatVu111 committed:
feat(backend): Add user signup, onboarding, and fetch logic

- Added , , and  endpoints.
- Updated  and  to handle user creation, updates, and fetching.
- Added  logic for tracking onboarding status.
- Updated database schema for datvu table.

ThanhDatVu111 committed:
Adjusted project setup to ensure compatibility with Expo Go and Xcode.

aanh1009 commented on frontend/src/app/(auth)/signin.tsx:
```diff
@@ -26,7 +26,7 @@ export default function Page() {
+        router.replace('../(tabs)/home');
       } else {
         console.error(JSON.stringify(signInAttempt, null, 2));
       }
```
after merging the pairing logic, we will need another API route to check if the person has been paired or not, if they have been paired, move them to the home tabs, if they're not, move them to the onboard page

ThanhDatVu111 replied:
Yes that is correct

ThanhDatVu111 committed:
Merge pull request #47 from ThanhDatVu111/bug/expo-go-version

Bug/expo go version

ThanhDatVu111 committed:
refactor: move user fetching logic to RootLayout for better routing

- Moved user fetching and routing logic from AuthRoutesLayout to RootLayout.
- Ensures users are redirected to the correct page (home or onboarding) upon app entry.

ThanhDatVu111 committed:
Refactor authentication screens for consistent UI and improved UX

tranguv commented on frontend/src/assets/icons/person.png:
You might already known this, but I'd like to introduce it to you anyway [Expo icon library](https://icons.expo.fyi/Index)

ThanhDatVu111 replied:
goddamn I did not know about this; thank you, miss

tranguv commented on backend/src/services/userService.ts:
```diff
@@ -6,9 +6,18 @@ export async function registerUser(input: any) {
   //Receive “raw” data from the controller and pass it to the model
```
Cast type any

ThanhDatVu111 replied:
Done

MC2114 reviewed (approved):
It works well on my machine, I did not run into ang bugs. The SDK version fix (in the `package.json`) will help with us viewing our product in Expo Go; and splitting the logic for updating users in the schema is necessary as our schema evolved.

ThanhDatVu111 committed:
just fix what copilot suggest

ThanhDatVu111 committed:
Merge pull request #49 from ThanhDatVu111/ux/auth-tabs

Refactor authentication screens for consistent UI and improved UX

Quinxie51 reviewed (approved):
lgtm

tranguv commented on frontend/src/app/_layout.tsx:
```diff
@@ -1,17 +1,56 @@
+import { useRouter } from 'expo-router';
+
+function AuthenticatedLayout() {
+  const { isLoaded, isSignedIn, userId } = useAuth();
```
Refactor this as a hook instead

ThanhDatVu111 replied:
Done

ThanhDatVu111 committed:
Extracted authentication and redirection logic into a reusable useAuthGuard hook.

ThanhDatVu111 committed:
Added specific TypeScript interfaces (RegisterUserInput, OnboardUserInput, FetchUserInput) to replace any types.
